reactor-core/src/main/java/reactor/core/publisher/OnNextFailureStrategy.java

```diff
@@ -0,0 +1,274 @@
+	}
+
+	@Override
+	boolean test(Throwable throwable, @Nullable Object o);
+
+	/**
+	 * Process an error and the optional value that caused it (when applicable) in
+	 * preparation for sequence resume, so that the error is not completely swallowed.
+	 * <p>
+	 * If the strategy cannot resume this kind of error (ie. {@link #test(Throwable, Object)}
+	 * returns false), return the original error. Any exception in the processing will be
+	 * caught and returned. If the strategy was able to process the error correctly,
+	 * returns null.
+	 *
+	 * @param error the error being recovered from.
+	 * @param value the value causing the error, null if not applicable.
+	 * @param context the {@link Context} associated with the recovering sequence.
+	 * @return null if the error was processed for resume, a {@link Throwable} to propagate
+	 * otherwise.
+	 * @see #test(Throwable, Object)
+	 */
+	@Nullable
+	Throwable process(Throwable error, @Nullable Object value, Context context);
+
+	/**
+	 * A strategy that never let any error resume.
+	 */
+	static OnNextFailureStrategy stop() {
+		return STOP;
+	}
+
+	/**
+	 * A strategy that let all error resume. When processing, the error is passed to the
+	 * {@link Operators#onErrorDropped(Throwable, Context)} hook and the incriminating
+	 * source value (if available) is passed to the {@link Operators#onNextDropped(Object, Context)}
+	 * hook, allowing the sequence to continue with further values.
+	 */
+	static OnNextFailureStrategy resumeDrop() {
+		return RESUME_DROP;
+	}
+
+	/**
+	 * A strategy that let some error resume. When processing, the error is passed to the
+	 * {@link Operators#onErrorDropped(Throwable, Context)} hook and the incriminating
+	 * source value (if available) is passed to the {@link Operators#onNextDropped(Object, Context)}
+	 * hook, allowing the sequence to continue with further values.
+	 * <p>
+	 * Any exception thrown by the predicate is thrown as is.
+	 *
+	 * @param causePredicate the predicate to match in order to resume from an error.
+	 */
+	static OnNextFailureStrategy resumeDropIf(Predicate<Throwable> causePredicate) {
+		return new ResumeDropStrategy(causePredicate);
+	}
+
+	/**
+	 * A strategy that let all errors resume. When processing, the error and the
+	 * incriminating source value are passed to custom {@link Consumer Consumers}.
+	 * <p>
+	 * Any exception thrown by the consumers will suppress the original error and be
+	 * returned for propagation. If the original error is fatal then it is thrown
+	 * upon processing it (see {@link Exceptions#throwIfFatal(Throwable)}).
+	 *
+	 *
+	 * @param errorConsumer the {@link Consumer} to process the recovered errors with.
+	 * @param valueConsumer the {@link Consumer} to process the error-causing values with.
+	 * It must deal with potential {@code null}s.
+	 * @return a new {@link OnNextFailureStrategy} that allows resuming the sequence.
+	 */
+	static OnNextFailureStrategy resume(Consumer<Throwable> errorConsumer,
```

Comment:
@simonbasle Why not a `BiConsumer` here? Especially given the two are so related given the error is caused by onNext of the value.

Reply:
That makes sense. I think at some point the dropping implementation was reusing the consumer implementation, so having separate consumers would allow to detect the `value == null` case and skip the value consumer entirely, making the reuse as simple as calling `Operators::onNextDropped`.

Reply:
Maybe, but then it's a significant issue if/when you need to use both even if it's a simple as log a message referencing error and value.  Any issue with switching this to BiConsumer, and doing the adaption to next dropped consumer internally?